[Navigation] Makes URL dynamic for pages and posts

draganescu opened the pull request:
## Description
Closes #18345

## How has this been tested?
Tested locally by:
1. Make sure you have some posts and pages
2. Add a new page
3. Add a Navigation block
4. Create empty
5. Add a link to a page, a link to a post and an external link
6. Save and view post
7. Links should be working

## Types of changes
Small change to `NavigationLinkEdit` to also save the `type` which `LinkControl` offers to the `onChange` handler.

Then on the render function (PHP) added a call to `get_permalink` for anything which is not an external URL.

draganescu committed:
adds simple dynamic link generation to navigation items

draganescu committed:
Linting and snapshots fix

draganescu commented:
This makes the URL dynamic for pages and posts on rendering the block but does not handle the URL stored as an attribute for pages and posts.

I wonder how this can be solved though? 

1) Make an API call for each menu item to get the latest URL? Seems overkill.
2) Make LinkControl get latest URL on edit for type page or post? Could make editing clunky.
3) Add a sort of `get_permalink` clientside? Without calling the API it's a duplicated func. to maintain. With calling the api is similar to 1 or 2.

Just to be clear, the problem is when we have a menu saved and settings change (say permalink structure) and someone edits a menu item, they should see the updated link.

cc @mtias

aduth commented on packages/block-library/src/navigation-link/edit.js:
```diff
@@ -278,6 +279,7 @@ function NavigationLinkEdit( {
 											return escape( normalizedURL );
 										} )(),
 										opensInNewTab: newOpensInNewTab,
+										type,
```
Where is this defined in the attribute schema of the block?

draganescu replied:
[Here](https://github.com/WordPress/gutenberg/blob/8e4d714c9da0ec979cbb14a16ee47d1a1fba0293/packages/block-library/src/navigation-link/block.json#L12) we have type.

aduth commented on packages/block-library/src/navigation-link/edit.js:
```diff
@@ -278,6 +279,7 @@ function NavigationLinkEdit( {
 											return escape( normalizedURL );
 										} )(),
 										opensInNewTab: newOpensInNewTab,
+										type,
```
These suggestion types are proposed to be removed by the issue at #20051 . Is it strictly necessary for them to exist in this implementation, in mind of the problems and alternatives considered in #20051 ?

draganescu replied:
It depends. I don't see a problem with having type as `LinkControl` does create many types of items which depend on the, well, type of suggestion picked by the user. In this context as you can see it is extremely useful to have it as it becomes easy to render dynamic links if the URL is not manually entered.

If we remove the type we'd check for id then. But it does seem like a good descriptor to have since we actually do have many types and we treat them differently. I am not a fan of inferring categories from data values when I can store a flat thing that tells me what it is.

aduth replied:
The problem for me is that these types are neither clearly documented nor based on some shared understanding of categorization of links. While they're not _wrong_, they feel arbitrary. My fear is that this will cause more confusion than it will help, both in that a consumer will not clearly understand which options are available to them (and how they relate to other "types" within WordPress like a post type), and that a future maintainer will not have confidence to make a decision in adding to or revising these types. Furthermore, if it is the case that these types can be derived based on other values (like `id`), then they can be considered redundant, which leads to additional problems of keeping values in sync when one changes.

I'd not mind if there were to be some utility functions which abstract this in a way which is more human-readable, but generally I think the further we can get with the fewest distinct attribute values, the better.

draganescu replied:
I removed the "type" checking and also fixed `LinkControl` to not set an id for manual entry URLs. :) \o/

talldan commented on packages/block-library/src/navigation/index.php:
```diff
@@ -209,10 +209,22 @@ function block_core_navigation_build_html( $attributes, $block, $colors, $font_s
 			$html .= ' href="' . esc_url( $block['attrs']['url'] ) . '"';
 		}
 
+		if (
```
It'd be good to avoid the possibility of two hrefs if another dev comes along later and makes a mistake with some of this logic. 

I thinkt the easiest option would be to use an `else` here.
```php
if (
  isset( $block['attrs']['url'] ) &&
  isset( $block['attrs']['type'] )
) {
    if ( 'URL' === $block['attrs']['type'] ) {
      // use the url attribute.
    } else {
      // get the permalink and use that.
    }
}
```

talldan commented on packages/block-library/src/navigation/index.php:
```diff
@@ -209,10 +209,22 @@ function block_core_navigation_build_html( $attributes, $block, $colors, $font_s
+			isset( $block['attrs']['id'] ) &&
+			'URL' !== $block['attrs']['type']
+		) {
+			$html .= ' href="' . get_permalink( $block['attrs']['id'] ) . '"';
```
Checking through core code, `get_permalink` seems to be often escaped using `esc_url` when used in markup though sometimes not.

I think best to stay on the side of caution and escape it.

draganescu committed:
removes checking by type

also fixes LinkControl to not set URL as if for manual entries

draganescu committed:
adds snapshot fix

draganescu committed:
removed URL as id from snapshot

draganescu committed:
fixes snapshot

draganescu commented:
This PR also adds to the Navigation block the small feat of rendering a `div` if the item is only a label and contains neither an ID or an URL.

talldan commented on packages/e2e-tests/specs/experiments/__snapshots__/navigation.test.js.snap:
```diff
@@ -2,19 +2,19 @@
 exports[`Navigation allows a navigation menu to be created using existing pages 1`] = `
 "<!-- wp:navigation {\\"orientation\\":\\"horizontal\\"} -->
-<!-- wp:navigation-link {\\"label\\":\\"Home\\",\\"type\\":\\"page\\",\\"id\\":1,\\"url\\":\\"https://this/is/a/test/page/home\\"} /-->
+<!-- wp:navigation-link {\\"label\\":\\"Home\\",\\"id\\":1,\\"url\\":\\"https://this/is/a/test/page/home\\"} /-->
```
If the link is dynamic, saving the `url` attribute seems unnecessary.

draganescu replied:
You are right.

talldan commented on packages/block-library/src/navigation/index.php:
```diff
@@ -212,16 +212,28 @@ function block_core_navigation_build_html( $attributes, $block, $colors, $font_s
+			$item_tag = 'a';
+		}
+
 		$html .= '<li class="' . esc_attr( $css_classes . ( $has_submenu ? ' has-child' : '' ) ) .
 			( $is_active ? ' current-menu-item' : '' ) . '"' . $style_attribute . '>' .
-			'<a class="wp-block-navigation-link__content"';
+			'<' . $item_tag . ' class="wp-block-navigation-link__content"';
 
-		// Start appending HTML attributes to anchor tag.
-		if ( isset( $block['attrs']['url'] ) ) {
-			$html .= ' href="' . esc_url( $block['attrs']['url'] ) . '"';
+		if (
+			isset( $block['attrs']['id'] )
+		) {
+			$html .= ' href="' .
+				esc_url( get_permalink( $block['attrs']['id'] ) ) .
+			'"';
+		} else {
+			if ( 'a' === $item_tag ) {
+				$html .= ' href="' . esc_url( $block['attrs']['url'] ) . '"';
+			}
 		}
 
-		if ( isset( $block['attrs']['opensInNewTab'] ) && true === $block['attrs']['opensInNewTab'] ) {
+		if ( 'a' === $item_tag && isset( $block['attrs']['opensInNewTab'] ) && true === $block['attrs']['opensInNewTab'] ) {
```
Not a showstopper, but I feel like some of this logic could be made more readable. For example extracting the logic for getting the url to a function and then checking for the existence of a url first would tidy up some of the resulting if statement logic:
```php
$url = get_navigation_link_url( $block['attrs'] );
$has_url = ! is_empty( $url );
$item_tag = $has_url ? 'a' : 'div';

// ...

if ( $has_url ) {
    $html .= ' href="' . $url . '"';
}

// ...
if ( $has_url && isset( $block['attrs']['opensInNewTab'] ) && true === $block['attrs']['opensInNewTab'] ) {
```

draganescu replied:
Yep, I've cleaned the code a bit as suggested, thanks!

talldan commented on packages/block-editor/src/components/link-control/index.js:
```diff
@@ -269,7 +269,6 @@ function LinkControl( {
 
 		return Promise.resolve( [
 			{
-				id: val,
```
I'm not totally clear on what this change does, any background on the need to remove this?

draganescu replied:
The manual entries have no `id` anymore so I removed it from this resolve.

draganescu committed:
refactores according to code review

adds a new get_navigation_link_url function to navigation link rendering
removes url from navigation links that have id

draganescu committed:
removes URL when creating from existing pages

also PHP linting

draganescu committed:
corrects docblock

draganescu committed:
adding stored URL back

removing the URL from the navigation item caused LinkControl to not load in edit mode when a link is present. it also messes the look of the navigation items.

It is rather weird that we store a potentially outdated link but we can address that separately.

draganescu commented:
@talldan I added back the link to the navigation item. Once the snapshots work I would suggest we address this stale url stored as attribute in a separate PR. For now the `LinkContol` component is not ready to manage links without URLs and also the `NavigationLinkEdit` component expects the items to have URLs for proper styling.

I am afraid that updating LinkControl for this case would make things complex in this context. What do you think?

draganescu committed:
revert snapshot for adding the URL back to navigation link

draganescu committed:
fixes typoin snapshot

talldan commented:
So just to be clear, as I don't think it's all that clear for those outside this conversation—the bug this introduces is as follows:
1. Create a navigation block with a navigation link pointing to a post and save
2. Change the permalink structure for posts
3. Check the URL on the front-end, observe that the URL for the post in the front-end matches the new permalink structure.
4. Check the URL in the editor, observe that the URL for the post in the editor is outdated and matches the old permalink structure.

The current behavior in `master` is that both 3 & 4 are broken. So while this technically makes things less broken, it does introduce an inconsistency for the block that wasn't there before.

draganescu commented:
Looking at the `LinkControl` component it seems possible to update it to work with Id instead of URL. However, in order to have the blocks update their own stored Urls to use the rendered version (that is to take into account permalink structure and config changes) is a serious task.

As far as the `LinkControl` goes,

- when the init `value` is not an `Url`, and we determine it is some `Id`, then we fetch the `Url` from the API and show that. This way, we could show for pages and posts the titles instead of links.

- when the init `value` is empty or we edit,  then `LinkControl` will simply don't pass `Url` to the `onChange` callback in the various contexts it is used (e.g. click on suggestion)

The problem here is that all the parent components implementing `LinkControl` will need their own front end rendering system.

Even if `LinkControl` autoupdates the Url on edit, the moment it is passed to a component who has no clue how Urls work, it will be optionally stale.

Assuming we only update the navigation items to not store any URL since the front will potentially render another one, we run into displaying links with no link, or invalid anchors styled to look like links.

Even if we solve it for one block, say we customize only the `NavigationItem - LinkControl` relationship at least the formatted links in `RichText` will still be in danger of being outdated.

cc @mtias what do you think of this?

earnjam commented:
So many of these types of issues are related to this insistence on storing nav menus as serialized block data. We have an existing menu data storage method in WordPress that already has problems like these solved. I still don't really understand the reasoning for abandoning it.

mtias commented on packages/block-library/src/navigation/index.php:
```diff
@@ -212,16 +228,22 @@ function block_core_navigation_build_html( $attributes, $block, $colors, $font_s
 		};
 
+		$item_url = get_navigation_link_url( $block['attrs'] );
+		$item_tag = $item_url ? 'a' : 'div';
```
Why isn't this a `span` instead?

draganescu replied:
Yep good question :) a span is more fit as an inline non-link element.

draganescu commented:
@mtias is #21932 related to this as well?

mtias commented:
> The problem here is that all the parent components implementing LinkControl will need their own front end rendering system.

I don't think `LinkControl` should be updated for all cases it's used. When adding an inline link, we are still bound to rendering the link as entered by the user and there's currently no expectations of such links, even if internal, being updated on permalink changes. We also don't handle server rendering inline links.

We should only handle the **Link** block as used in **Navigation** for now. I'd imagine `LinkControl` could have a property to make it dynamic or to prioritize IDs.

> So many of these types of issues are related to this insistence on storing nav menus as serialized block data. We have an existing menu data storage method in WordPress that already has problems like these solved. I still don't really understand the reasoning for abandoning it.

@earnjam I think the issue is that the current setup addresses some issues but leaves quite a few others open. Once we start allowing more blocks to be inserted in menus — search, social icons, headings, images, text — the storage breaks down a bit as it needs to handle many other representations, some dynamic and some static. Depending on how the design evolves, we might need to have specialized blocks for any set of links (like sub-pages) so that they can be better intermixed with other blocks.

In any case, the navigation block currently works with both data structures depending on where it's used (site editing or nav-menus.php screen) so we can see where the different issues arise.

Also worth emphasizing for this particular thread that the **Link** block is not necessarily attached to navigation, it could become a standalone block for various other uses, so it still needs to model the attributes properly.

draganescu committed:
updates LinkControl to work with only id instead of URL

draganescu committed:
removes link from predefined navigation options, fixes create from menu bug

draganescu commented:
I have updated this PR to never save the URL in Link blocks and the LinkControl component to display only a label when the link is containing only an ID but no URL.

@talldan can you take this for a new spin?

draganescu committed:
fixes tests

draganescu committed:
snapshot update

draganescu committed:
fix for snapshot

talldan commented on packages/block-editor/src/components/link-control/index.js:
```diff
@@ -158,6 +158,7 @@ const makeCancelable = ( promise ) => {
 function LinkControl( {
 	searchInputPlaceholder,
 	value,
+	id,
```
I'm not sure about introducing `id` as a separate prop. It's really just introducing a second `value`, but there's already a `value` prop that should serve this purpose.

It feels like instead the Navigation Link block should be able to omit the `url` from the `value` prop, instead defining the `id` to achieve the same thing:
```js
{
    id: 1
}
```

and for custom links it could omit the `id`, instead specifying the `url`:
```js
{
    url: 'https://google.com',
}
```

talldan reviewed (commented):
I'm not completely comfortable about these changes to LinkControl.

I think we have to be pretty careful about making changes to it, given how difficult URLInput became to work with.

draganescu committed:
revert for rebase

draganescu committed:
updates for the new self rendering link component

draganescu committed:
fixed for get_navigation_link_url function rename

draganescu committed:
removes id from LinkControl

draganescu committed:
refactors un-needed function wrapper

draganescu commented:
@talldan can you check this again, I added your suggestion to pass the id as part of value, good idea! Thank you!

adamziel commented on packages/block-library/src/navigation-link/index.php:
```diff
@@ -171,7 +194,7 @@ function render_block_core_navigation_link( $attributes, $content, $block ) {
+	$html .= '</' . $item_tag . '>';
 	// End anchor tag content.
 
 	if ( $block->context['showSubmenuIcon'] && $has_submenu ) {
```
Not specific to this PR but sometimes it generates a notice.

```suggestion
	if ( isset( $block->context['showSubmenuIcon'] ) && $block->context['showSubmenuIcon'] && $has_submenu ) {
```

adamziel commented:
Here's what I see in the post editor after applying this PR:

![Uploading Zrzut ekranu 2020-07-14 o 16.09.03.png…]()

Interestingly, the block is rendered on the frontend without any issues. I wonder if there's something wrong with my working copy? Hopefully an error like that would trigger some test failures but they seem to be in tact.

adamziel commented:
And it fixed itself, super interesting!

draganescu committed:
Update packages/block-library/src/navigation-link/index.php

Co-authored-by: Adam Zielinski <adam@adamziel.com>

adamziel commented:
Before:
<img width="495" alt="Zrzut ekranu 2020-07-14 o 16 35 33" src="https://user-images.githubusercontent.com/205419/87444490-26fe1080-c5f7-11ea-91b4-110eebe47a6e.png">
<img width="383" alt="Zrzut ekranu 2020-07-14 o 17 26 57" src="https://user-images.githubusercontent.com/205419/87444586-4301b200-c5f7-11ea-951b-d2d925988189.png">


After (with permalinks enabled):
<img width="392" alt="Zrzut ekranu 2020-07-14 o 16 18 42" src="https://user-images.githubusercontent.com/205419/87444486-25cce380-c5f7-11ea-84e9-114fcb71c2b8.png">
<img width="395" alt="Zrzut ekranu 2020-07-14 o 17 38 12" src="https://user-images.githubusercontent.com/205419/87445826-d38cc200-c5f8-11ea-951b-8a7573a52b82.png">

Two problems here - I don't know what the link is anymore, and after clicking edit the input is empty.

adamziel commented:
It's fine to update the data model so that it tracks the ID instead of the URL, but the editor experience should make it seamless to the user. I don't like how you're not able to preview (or edit) the URL anymore. Maybe we should fetch all the relevant URL once the editor is loaded?

draganescu commented:
To enable that kind of editing we should fetch from the API details about the linked page. We don't know ahead of time which are the "relevant" pages so we'll end up with a spinner in that popover, so we can know the URL for the id. So it's still a downside. Also, when you edit the link, the link control will have to revert to "manual link" loosing the connection to the id.

talldan commented on packages/block-library/src/navigation-link/edit.js:
```diff
@@ -60,10 +60,12 @@ function NavigationLinkEdit( {
 		url,
+		id,
 		opensInNewTab,
+		title: label,
```
Hmm, these two aren't really the same thing. If I edit the label it doesn't mean that my page title is changed.

Not sure what the alternative is, since we don't actually know the page title.

talldan commented on packages/block-library/src/navigation-link/index.php:
```diff
@@ -171,10 +194,10 @@ function render_block_core_navigation_link( $attributes, $content, $block ) {
 	// End anchor tag content.
 
-	if ( $block->context['showSubmenuIcon'] && $has_submenu ) {
+	if ( isset( $block->context['showSubmenuIcon'] ) && $block->context['showSubmenuIcon'] && $has_submenu ) {
```
This change seems unrelated. Is it needed?

draganescu replied:
It was a suggestion from @adamziel to make sure we don't get PHP Notices sometimes.

talldan commented:
> To enable that kind of editing we should fetch from the API details about the linked page. We don't know ahead of time which are the "relevant" pages so we'll end up with a spinner in that popover, so we can know the URL for the id. So it's still a downside. Also, when you edit the link, the link control will have to revert to "manual link" loosing the connection to the id.

I think when you edit, the page title should be the value in the input box.

adamziel commented:
> To enable that kind of editing we should fetch from the API details about the linked page. We don't know ahead of time which are the "relevant" pages so we'll end up with a spinner in that popover, so we can know the URL for the id. 

How I imagine it is this:

* When the NavigationLink is first mounted, it requests it's related page from the API. A more sophisticated approach could batch these requests too.
* When a page is selected from the suggestion box, the URL is retained in state.

> So it's still a downside. Also, when you edit the link, the link control will have to revert to "manual link" loosing the connection to the id.

That's one way to approach it - it sounds fine for me, but maybe we could figure out even better UI here, for example disallow editing the page URL while still making it possible to edit the link URL:

<img width="557" alt="Zrzut ekranu 2020-07-15 o 12 37 12" src="https://user-images.githubusercontent.com/205419/87535638-1a7dc480-c698-11ea-9d51-d0dd0c87cfcb.png">

^ This isn't the right solution, but my thinking goes in that direction. Also CC @shaunandrews since it's super relevant for in-toolbar editing.

draganescu commented:
@adamziel do you think #24099 kinda blocks this? I do!

noisysocks commented:
Closing this as it's gotten stale. I'll try to address #18345 in a separate PR. It's closely related to what I've been working on in https://github.com/WordPress/gutenberg/pull/24670.
